pkg/infrastructure/clusterapi/clusterapi.go

```diff
@@ -0,0 +1,279 @@
+package clusterapi
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	utilkubeconfig "sigs.k8s.io/cluster-api/util/kubeconfig"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/cluster/metadata"
+	"github.com/openshift/installer/pkg/asset/ignition/bootstrap"
+	"github.com/openshift/installer/pkg/asset/ignition/machine"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	"github.com/openshift/installer/pkg/asset/kubeconfig"
+	"github.com/openshift/installer/pkg/asset/manifests/capiutils"
+	capimanifests "github.com/openshift/installer/pkg/asset/manifests/clusterapi"
+	"github.com/openshift/installer/pkg/clusterapi"
+	"github.com/openshift/installer/pkg/infrastructure"
+	"github.com/openshift/installer/pkg/types"
+)
+
+// InfraProvider is the base implementation for provisioning cluster
+// infrastructure using CAPI. Platforms should embed this struct and
+// implement:
+// .
+type InfraProvider struct {
+	CAPIInfraHelper
+}
+
+// InitializeProvider returns a CAPI provider implementation for a specific
+// cloud platform.
+func InitializeProvider(platform CAPIInfraHelper) infrastructure.Provider {
+	return InfraProvider{platform}
+}
+
+// CAPIInfraHelper provides an interface for calling functions at different
+// points of the CAPI infrastructure provisioning lifecycle.
+type CAPIInfraHelper interface {
+	// PreProvision is called before provisioning using CAPI controllers has begun.
+	// and should be used to create dependencies needed for CAPI provisioning,
+	// such as IAM roles or policies.
+	PreProvision(in PreProvisionInput) error
+
+	// ControlPlaneAvailable is called once cluster.Spec.ControlPlaneEndpoint.IsValid()
+	// returns true, typically after load balancers have been provisioned. It can be used
+	// to create DNS records.
+	ControlPlaneAvailable(in ControlPlaneAvailableInput) error
+}
```

Comment:
As a side note, having all of them comprised in a single interface may be handy from a maintenance point of view, but will force the user to implement all of them - even those ones that are not really required. I think it's ok unless this interface is meant to grow with an excessive number of methods

Reply:
This suggestion was mentioned a few times in the enhancement, but I am missing how decomposing the interface would allow some implementers to skip implementing the unneeded function. Instead, I have created a default implementation that other providers can embed. This should have the same effect of not requiring each provider to implement all (or even any) functions: 726fe329810bb1c86f7f0c6209c8b2c25f7a9f15